### PR TITLE
ETL Data Structures: NewAccumulation()

### PR DESCRIPTION
**Pull request by mbolt35**
## What does this PR change?
* Adds a `NewAccumulation()` method on the `SetRange` ETL data structures which provides an Accumulate method which does not mutate. 

## How will this PR impact users?
* Fixes a caching bug 

## How was this PR tested?
* Unit Tests and via endpoint

### CODE REVIEW

File: pkg/kubecost/asset.go [hunk 1]
```diff
@@ -68,63 +68,65 @@ type Asset interface {
...
+//     => Allocation{Name: "monitoring/__unallocated__", ExternalCost: 10.00, TotalCost: 10.00}, nil
 //
-//   4) no match
-//   aggregateBy = ["cluster"]
-//   allocationPropertyLabels = {"namespace":"kubernetes_namespace"}
-//   => nil, err
+//  4. no match
+//     aggregateBy = ["cluster"]
+//     allocationPropertyLabels = {"namespace":"kubernetes_namespace"}
+//     => nil, err
```
Comment (mbolt35) [inline pkg/kubecost/asset.go:129]: Apparently I have a VSCode extension that organizes comments (I sort of like it, although I didn't expect these changes).

**Review by michaelmdresser — APPROVED** [review]
LGTM! Commit history looks a little funky, though.

File: pkg/kubecost/allocation.go [hunk 2]
```diff
@@ -1988,6 +1989,36 @@ func (asr *AllocationSetRange) Accumulate() (*AllocationSet, error) {
 	return allocSet, nil
 }
 
+// NewAccumulation clones the first available AllocationSet to use as the data structure to
+// accumulate the remaining data. This leaves the original AllocationSetRange intact.
+func (asr *AllocationSetRange) NewAccumulation() (*AllocationSet, error) {
```
Comment (michaelmdresser) [inline pkg/kubecost/allocation.go:1994]: Now that I've thought about it some more, is there any reason this shouldn't be the logic for the standard `Accumulate()`?

**Comment by mbolt35** [comment]
> LGTM! Commit history looks a little funky, though.

Yeah, it's because I forgot to commit a sig, then accidentally rebased my sig 1 commit too far in the past. Just going to squash on merge.

File: pkg/kubecost/allocation.go (hunk 2 above)
Comment (mbolt35) [reply pkg/kubecost/allocation.go:1994]: So, for regular Allocations, from my testing, `NewAccumulation()` almost works exactly like `Accumulate()`, but this is not the case for `AssetSetRange` or `SummaryAllocationSetRange`. For API consistency, the `New___` variation seemed more explicit in that it created copies of the inner data before squashing. I _did_ explore updating `AllocationSetRange::Accumulate()` to match `AssetSetRange` and `SummaryAllocationSetRange`, but there are a large handful of tests which fail after that change, which seems to indicate an expected behavior there that's different from the latter two. 

For the most part, `Accumulate()` is fine to run given that the final SetRange object won't be used post accumulation. However, in the case with caching, we want the raw SetRange to remain immutable, so we need an immutable method _if_ accumulate=true.

Comment (michaelmdresser) [reply pkg/kubecost/allocation.go:1994]: Sounds perfect, thanks for the thorough response. `New___` is perfect here, I'm a big fan.

File: pkg/kubecost/asset.go (hunk 1 above)
Comment (nikovacevic) [reply pkg/kubecost/asset.go:129]: Hahaha

**Review by nikovacevic — APPROVED** [review]
Thanks for doing this Bolt. Dig it. We should probably universalize Accumulate around the way we like it, I agree; but this is wonderful.

**Pull request merged**